winch/codegen/src/codegen/mod.rs

```diff
@@ -101,7 +182,13 @@ where
 
         while !body.eof() {
             let offset = body.original_position();
-            body.visit_operator(&mut ValidateThenVisit(validator.visitor(offset), self))??;
+            if self.context.reachable {
+                body.visit_operator(&mut ValidateThenVisit(validator.visitor(offset), self))??;
+            } else {
+                let op = body.read_operator()?;
+                validator.op(offset, &op)?;
+                self.handle_unreachable_state(&op);
+            }
```

Comment:
Bit of a drive-by comment, but altering the core loop here felt a bit unfortunate mainly along the lines of the translation of `If` for example was spread across `handle_unreachable_state` as well as the typical visitor.

One option though might be to use macro-magic to get `ValidateThenVisit` to skip all instructions except the allow-listed control flow instructions when the state is `unreachable`, sort of like how the visitor today fails on all instructions except for an allow-listed set that's implemented?

Reply:
If the macro approach doesn't work and if there's a strong preference to avoid mixing the visitors with match-on-op, @cfallin's suggestion is what I had in mind as a fallback. One thing that I like of the current approach, is that the handling of unreachable code is isolated and should only be updated if there's a need to deal with such state while handling the unreachable code state in each visitor function is a bit more manual and error prone, but in the spirit of more explicit code I think that's fine.

Reply:
Also to be clear I don't feel strongly about this at all, so whatever feels like it works best I think is fine